### DISCUSSION TITLE
Making a TEC/Peltier library from UdSAES work.

### DISCUSSION THREAD

**Discussion by ryanwhite-stemcell** [OpenModelica/OpenModelica#11191] [General]
I'm trying to model a system that uses TECs, and I found this library from this old forum post: https://openmodelica.org/forum/default-topic/2325-i-am-looking-for-a-modelica-model-of-a-thermoelectric-module 

https://www.uni-saarland.de/fileadmin/upload/lehrstuhl/frey/software/ThermoelectricGenerator_V5_7.zip

It seems to be from https://github.com/UdSAES but not on their GitHub profile...not sure what's up there. 

Anyways, the examples hang with 
```
[5] 14:40:58 Translation Notification
[ThermoelectricGenerator_V5_7: 7992:35-7992:47]: [From here:](omeditmessagesbrowser:///ThermoelectricGenerator_V5_7?lineNumber=7992)

[6] 14:40:58 Translation Error
[ThermoelectricGenerator_V5_7: 7992:7-7993:66]: [Non-array modification ‘"Ohm.m"‘ for array component ‘unit‘, possibly due to missing ‘each‘.](omeditmessagesbrowser:///ThermoelectricGenerator_V5_7?lineNumber=7992)
```

The offending line looks like this: 
```      
parameter Real rho_el[:, :](unit="Ohm.m") = {{0,0.01}}
        "Electrical resistivity" annotation (dialog(enable=true));
```
If I change it to `parameter Real rho_el[:, :](each unit="Ohm.m") = {{0,0.01}}`

I get: 
```
[5] 14:42:27 Translation Error
[ThermoelectricGenerator_V5_7: 6433:11-6437:21]: Type mismatch for positional argument 4 in ThermoelectricGenerator_V5_7.Inner_TEG.Components_for_TEL.Base_Components.Functions.tableInit(smoothness=1). The argument has type:
  Integer
expected type:
  enumeration Modelica.Blocks.Types.Smoothness(LinearSegments, ContinuousDerivative, ConstantSegments, MonotoneContinuousDerivative1, MonotoneContinuousDerivative2)
```

I feel like I'm going down a rabbit hole with this, but I'd love to make it work. Any tips?

**Comment by casella** [comment]
The rabbit hole is hopefully not that deep, but it's hard to say...

There are some hints here: https://trac.openmodelica.org/OpenModelica/wiki/WritingCompliantLibraries

Specifically, some libraries rely on the fact that Dymola accepts integers in lieu of enumerations. This is actually not allowed by the Modelica language specification, so we don't support that.

Good luck!

**Reply by ryanwhite-stemcell** [reply]
Ahh excellent. Thanks so much for that. I'm sure that's what's up. I have a trial version of Dymola but unfortunately everything is too complex for its library size limitations, and it doesn't include any useful library licenses...!

I'll keep digging.

**Comment by casella** [comment]
If you manage to get this library to work, it would be nice to post it on GitHub, so others can use it with omc.

**Comment by ryanwhite-stemcell** [comment]
I think I got it working. In all the blocks like 
```
          ID_Int_alpha = Base_Components.Functions.tableInit(
                  dummy,
                  dummy,
                  Material.Int_alpha,
                  1);
``` 

I changed the `1` to `Modelica.Blocks.Types.Smoothness.ContinuousDerivative`

I also left in that `parameter Real rho_el[:, :](each unit="Ohm.m") = {{0,0.01}}` change. 

At least things now seem to simulate. Whether the simulations are valid or not....that's a problem for Monday!
